// @microsoft/mgt-components/dist/es6/components/mgt-person/mgt-person-types.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
/**
 * Enumeration to define what parts of the person component render
 *
 * @export
 * @enum {number}
 */
export var PersonViewType;
(function (PersonViewType) {
    /**
     * Render only the avatar
     */
    PersonViewType[PersonViewType["avatar"] = 2] = "avatar";
    /**
     * Render the avatar and one line of text
     */
    PersonViewType[PersonViewType["oneline"] = 3] = "oneline";
    /**
     * Render the avatar and two lines of text
     */
    PersonViewType[PersonViewType["twolines"] = 4] = "twolines";
    /**
     * Render the avatar and three lines of text
     */
    PersonViewType[PersonViewType["threelines"] = 5] = "threelines";
    /**
     * Render the avatar and four lines of text
     */
    PersonViewType[PersonViewType["fourlines"] = 6] = "fourlines";
})(PersonViewType || (PersonViewType = {}));
export var avatarType;
(function (avatarType) {
    /**
     * Renders avatar photo if available, falls back to initials
     */
    avatarType["photo"] = "photo";
    /**
     * Forces render avatar initials
     */
    avatarType["initials"] = "initials";
})(avatarType || (avatarType = {}));
