// @microsoft/mgt-components/dist/es6/components/sub-components/mgt-dot-options/mgt-dot-options.js
/**
 * -------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
 * See License in the project root for license information.
 * -------------------------------------------------------------------------------------------
 */
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { fluentMenu, fluentMenuItem, fluentButton } from '@fluentui/web-components';
import { MgtBaseComponent, customElement } from '@microsoft/mgt-element';
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { strings } from './strings';
import { registerFluentComponents } from '../../../utils/FluentComponents';
import { styles } from './mgt-dot-options-css';
registerFluentComponents(fluentMenu, fluentMenuItem, fluentButton);
/**
 * Custom Component used to handle an arrow rendering for TaskGroups utilized in the task component.
 *
 * @export MgtDotOptions
 * @class MgtDotOptions
 * @extends {MgtBaseComponent}
 */
let MgtDotOptions = class MgtDotOptions extends MgtBaseComponent {
    constructor() {
        super(...arguments);
        this._clickHandler = () => (this.open = false);
        this.handleItemClick = (e, fn) => {
            e.preventDefault();
            e.stopPropagation();
            fn(e);
            this.open = false;
        };
        this.handleItemKeydown = (e, fn) => {
            this.handleKeydownMenuOption(e);
            fn(e);
            this.open = false;
        };
        this.onDotClick = (e) => {
            e.preventDefault();
            e.stopPropagation();
            this.open = !this.open;
        };
        this.onDotKeydown = (e) => {
            if (e.key === 'Enter') {
                e.preventDefault();
                e.stopPropagation();
                this.open = !this.open;
            }
        };
    }
    /**
     * Array of styles to apply to the element. The styles should be defined
     * user the `css` tag function.
     */
    static get styles() {
        return styles;
    }
    /**
     * Strings for localization
     *
     * @readonly
     * @protected
     * @memberof MgtDotOptions
     */
    get strings() {
        return strings;
    }
    connectedCallback() {
        super.connectedCallback();
        window.addEventListener('click', this._clickHandler);
    }
    disconnectedCallback() {
        window.removeEventListener('click', this._clickHandler);
        super.disconnectedCallback();
    }
    /**
     * Invoked on each update to perform rendering tasks. This method must return
     * a lit-html TemplateResult. Setting properties inside this method will *not*
     * trigger the element to update.
     */
    render() {
        const menuOptions = Object.keys(this.options);
        return html(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      <fluent-button\n        appearance=\"stealth\"\n        aria-label=", "\n        @click=", "\n        @keydown=", "\n        class=\"dot-icon\">\uE712</fluent-button>\n      <fluent-menu class=", ">\n        ", "\n      </fluent-menu>"], ["\n      <fluent-button\n        appearance=\"stealth\"\n        aria-label=", "\n        @click=", "\n        @keydown=", "\n        class=\"dot-icon\">\\uE712</fluent-button>\n      <fluent-menu class=", ">\n        ", "\n      </fluent-menu>"])), this.strings.dotOptionsTitle, this.onDotClick, this.onDotKeydown, classMap({ menu: true, open: this.open }), menuOptions.map(opt => this.getMenuOption(opt, this.options[opt])));
    }
    /**
     * Used by the render method to attach click handler to each dot item
     *
     * @param {string} name
     * @param {MenuOptionEventFunction} clickFn
     * @returns
     * @memberof MgtDotOptions
     */
    getMenuOption(name, clickFn) {
        return html `
      <fluent-menu-item
        @click=${(e) => this.handleItemClick(e, clickFn)}
        @keydown=${(e) => this.handleItemKeydown(e, clickFn)}>
          ${name}
      </fluent-menu-item>`;
    }
    handleKeydownMenuOption(e) {
        if (e.key === 'Enter') {
            e.preventDefault();
            e.stopPropagation();
        }
    }
};
__decorate([
    property({ type: Boolean }),
    __metadata("design:type", Boolean)
], MgtDotOptions.prototype, "open", void 0);
__decorate([
    property({ type: Object }),
    __metadata("design:type", Object)
], MgtDotOptions.prototype, "options", void 0);
MgtDotOptions = __decorate([
    customElement('dot-options')
], MgtDotOptions);
export { MgtDotOptions };
var templateObject_1;
